nni/mutable/utils.py

```diff
@@ -0,0 +1,207 @@
+from typing import Any
+
+_last_uid = defaultdict(int)
+
+_logger = logging.getLogger(__name__)
+
+
+def uid(namespace: str = 'default') -> int:
+    """Global counter for unique id. Not thread-safe."""
+    _last_uid[namespace] += 1
+    return _last_uid[namespace]
+
+
+def reset_uid(namespace: str = 'default') -> None:
+    """Reset counter for a specific namespace."""
+    _last_uid[namespace] = 0
+
+
+class NoContextError(Exception):
+    """Exception raised when context is missing."""
+    pass
+
+
+class ContextStack:
+    """
+    This is to maintain a globally-accessible context environment that is visible to everywhere.
+
+    To initiate::
+
+        with ContextStack(namespace, value):
+            ...
+
+    Inside the context, you can access the nearest value put into ``with``::
+
+        get_current_context(namespace)
+
+    Notes
+    -----
+    :class:`ContextStack` is not multi-processing safe. Also, the values will get cleared for a new process.
+    """
+
+    _stack: dict[str, list] = defaultdict(list)
+
+    def __init__(self, key: str, value: Any):
+        self.key = key
+        self.value = value
+
+    def __enter__(self):
+        self.push(self.key, self.value)
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.pop(self.key)
+
+    @classmethod
+    def push(cls, key: str, value: Any):
+        cls._stack[key].append(value)
+
+    @classmethod
+    def pop(cls, key: str) -> Any:
+        if not cls._stack[key]:
+            raise NoContextError(f'Context with key {key} is empty.')
+        return cls._stack[key].pop()
+
+    @classmethod
+    def top(cls, key: str) -> Any:
+        if not cls._stack[key]:
+            raise NoContextError(f'Context with key {key} is empty.')
+        return cls._stack[key][-1]
+
+    @classmethod
+    def stack(cls, key: str) -> list:
+        return list(cls._stack[key])
+
+
+def get_current_context(key: str) -> Any:
+    return ContextStack.top(key)
+
+
+# The default namespace
+_DEFAULT_LABEL_NAMESPACE = 'param'
+
+
+class LabelNamespace:
+    """
+    To support automatic labeling of mutables.
+
+    When entering a :class:`LabelNamespace` context, the number of mutables will recount from 1,
+    and thus the generation of labels are reproducible.
+
+    The label namespace is NOT thread-safe. The behavior is undefined if multiple threads are
+    trying to enter the namespace at the same time.
+
+    A namespace is bounded to a key. Namespace bounded to different keys are completed isolated.
+    Namespace can have sub-namespaces (with the same key). The numbering will be chained (e.g., ``model_1_4_2``).
+
+    :class:`LabelNamespace` is implemented based on :class:`ContextStack`.
+
+    Examples
+    --------
+    >>> with LabelNamespace('param'):
+    ...     label1 = auto_label()       # param_1
+    ...     label2 = auto_label()       # param_2
+    ...     with LabelNamespace('param'):
+    ...         label3 = auto_label()   # param_1_1
```

Comment:
why is it `param_1_1` not `param_2_1`?

Reply:
I think the docstring is out-of-date. Just updated the docs and corresponding tests.